### DISCUSSION TITLE
Update image in for loop

### DISCUSSION THREAD

**Discussion by anderssonf** [fastplotlib/fastplotlib#417] [Q&A] [answered]
Is it possible to update an image in a for loop? Something similar to drawnow in matlab?

**Comment by kushalkolar** [comment]
Yes, an animation function, this applies for any graphic. See "Image updates" in the `simple.ipynb` example:

```python
# create a `Plot` instance
plot_v = Plot()

# make some random data
data = np.random.rand(512, 512)

# plot the data
plot_v.add_image(data=data, name="random-image")

# a function to update the image_graphic
# a plot will pass its plot instance to the animation function as an argument
def update_data(plot_instance):
    new_data = np.random.rand(512, 512)
    plot_instance["random-image"].data = new_data

#add this as an animation function
plot_v.add_animations(update_data)

# show the plot
plot_v.show()
```

**Reply by anderssonf** [reply]
Thank you Kushal!

I did read this example, but I did not understand how to modify the "new_data" from outside the function (update_date) in the for loop.

**Reply by kushalkolar** [reply]
fastplotlib is non-blocking, you can modify the data anytime after the graphc is created.

**Reply by kushalkolar** [reply]
It doesn't need to be within that function, you can have a for-loop in the global scope and modify the ImageGraphic data in the for-loop, but then the for-loop will probably block the renderer.

You could make an animation function that reads a global counter and updates the image based on that? If you provide context I can probably help with example code.

**Reply by anderssonf** [reply]
I tried to modify the example. I am trying to do something like this:

import numpy as np
from fastplotlib import Plot
from time import sleep
plot_v = Plot()
data = np.random.rand(512, 512)
plot_v.add_image(data=data, name="random-image")
plot_v.show(sidecar=True)

for j in range(1,10):
  plot_v["random-image"].data=np.random.rand(512, 512)
  print(j)
  sleep(1)

**Reply by kushalkolar** [reply]
yea that's going to block the renderer, maybe something like this could work (untested):

```python
# a global object for setting thedata
data = np.random.rand(512, 512)

def poll_data():
  global data
  plot["image"].data = data

plot.add_animations(poll_data)

for i in range(100):
  data = np.random.rand(512, 512)
```

However if you're really using a for-loop in the global scope to update an image I would probably revisit the design of what you're trying to do. Animations functions (added using `plot.add_animations`) are specifically used for running a function before or after rendering a frame. What is it that you're really trying to do?

**Reply by anderssonf** [reply]
I was unable to get the example to run. I guess that there should be a show() somewhere?

I am running some simulations and I would like to see what is happening during the simulations.

**Reply by kushalkolar** [reply]
Yup that code is just the relevant part for your question, you'll need to fit it into the rest.

I see, then yes a polling function that "taps into" your simulation data is probably the best way to do this. If the image data are large you include a `if` statement that checks if the data has changed, sinced updating very large arrays (such as 8k image) isn't as fast as a 512x512 image.

**Reply by anderssonf** [reply]
Apologies Kushal, but I don't understand how to fill in the rest. If I run the script below (with the for loop in a separate cell) the plot is just updated in the very end and not during the loop. Should there be a render/show/flush call somewhere?  

```python
from fastplotlib import Plot
import numpy as np
# a global object for setting thedata
data = np.random.rand(512, 512)
plot = Plot()
plot.add_image(data=data, name="image")
def poll_data():
  global data
  plot["image"].data = data

plot.add_animations(poll_data)

plot.show(sidecar=True)  

for i in range(100):
  data = np.random.rand(512, 512)
```

**Reply by kushalkolar** [reply]
I see what you mean now. I think the rendering engine, [`pygfx`](https://github.com/pygfx/pygfx), uses asyncio to handle the pre-render functions. I'd need to dig in more to figure out how it can be used with for loops. If you know about async functions and asyncio you might be able to figure it out.

You could also post at `pygfx` and ask how to modify the `Buffer` for a `WorldObject` in a for-loop, since `fastplotlib`  basically abstracts away the buffers and world objects into "plot items" such as images. When you do `plot["image"].data = <array>` it is basically modifying the buffer for an `Image` `WorldObject` and notifying that the buffer data should be uploaded to the GPU.

**Reply by anderssonf** [reply]
Thank you very much for the help! It would be very useful to be able to plot while computing.

**Reply by anderssonf** [reply]
Is it possible to use an offscreen canvas in fastplotlib? See
https://github.com/pygfx/wgpu-py/issues/471#issuecomment-1976411917

**Reply by kushalkolar** [reply, marked as answer]
I think the easiest way for you to do this right now is using two separate processes and sending the data to be plotted over zmq: https://github.com/fastplotlib/fastplotlib/tree/main/examples/notebooks/multiprocessing_zmq
